### PR TITLE
[IMP] website_event: improve web tracking

### PR DESCRIPTION
**Pull request by pyduf**
PURPOSE

When opening the event page on the website, there is a service worker in place
that pre-fetches all links in order to be able to work offline.
This commit aims to disable the page tracking for those prefetch requests to
avoid polluting the statistics with "fake" page views.

SPECS

- Add a preparation commit to introduce a "X-Disable-Tracking" header
- Put the header on all requests in the event module that prefetch content

See underlying commit for details.

Task-2476513

### CODE REVIEW

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pparidans) [inline addons/website_event_track/static/src/js/service_worker.js:309]: This shouldn't be altered. `pendingRequests` (aka "syncable requests") refers to requests that will introduce a change on the server-side and then shouldn't be considered as "prefetch" requests.

Note: In this case, only the "toggle reminder" route is concerned.

PS: I agree, there is missing doc here...

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pparidans) [inline addons/website_event_track/static/src/js/service_worker.js:253]: Please, adapt the documentation when changing the function's interface.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pparidans) [inline addons/website_event_track/static/src/js/service_worker.js:158]: Styling tip: you can avoid the key/val naming redundancy when the key == value's variable:

```suggestion
    return fetch(request, { headers });
```

**Review by pparidans — CHANGES_REQUESTED** [review]
The HTTP header based approach seems fine to me.

I leave it to the @odoo/rd-website team for the final word.

File: addons/website/models/ir_http.py
Comment (rdeodoo) [inline addons/website/models/ir_http.py:136]: Nitpicking: What do you think about a more generic name for this custom http header? In case we have other flows later not related to prefetching for which we don't want to track?

File: addons/website_event/views/event_templates_list.xml
Comment (rdeodoo) [inline addons/website_event/views/event_templates_list.xml:5]: @pierre-pde Maybe it is time to have a global check on all the controllers and pages and see if everything is tracked as we want to. For instance neither /blog or /shop is tracked right now, while it would be useful?
If you think those need to be enabled, you can just create another task so we don't noise this one.

**Review by rdeodoo — APPROVED** [review]
I didn't know about that offline PWA capability of the website_event module, it explains that weird tracking behavior.
Advanced feature but cool :o.

Anyway, LGTM, not much to say except if you can implement that as a fix in stable somehow that would be cool.
If not, it is fine as it does not really qualifies as a bug :+1:

File: addons/website_event/views/event_templates_list.xml
Comment (pierre-pde) [reply addons/website_event/views/event_templates_list.xml:5]: Hi @rdeodoo , as discussed, let's see what can be done. 
We actually have the opportunity to have something great here.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js]: Thanks for the information, I don't know much about service workers. I have changed and it still works.

Notes: 
- I see that this function is called from an event listener "sync". No risk that the page stored in the cache be refreshed through this listener ? (then the page will be tracked while the user hasn't access the page)
- following a remark below, I have instead of X-Prefetch header, used X-Do-Not-Track header

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js]: Yes of course, it is done

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js]: Thanks for the tip; it is changed.

File: addons/website/models/ir_http.py
Comment (pyduf) [reply addons/website/models/ir_http.py]: Yes, thanks for the remark, it has the advantage of simplicity, it conveys directly the goal of the header. I have changed X-Prefetch for X-Do-Not-track.

**Comment by rdeodoo** [comment]
It was just a suggestion but if you consider it it's cool.
> Maybe we could continue on master, then after merge see to backport part of it 

I guess I'd go for master too to keep it simple here and see afterward :+1: 

> as change seems backward compatible

Can probably be done indeed without the signature change? I'll let you decide if that's really something you want to spend time on it to ship it in stable or not.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js]: As mentioned above, I don't know much about service worker so I have a newbie question: 
Will the method processPendingRequests called by a listener "sync" not fetch URL not seen by the user (by refreshing the cache) ?
In that case I should add the X-Do-Not-Track header (not X-Prefetch anymore as explained above) to prevent the page being wrongly tracked.

Thanks for your advice

File: addons/website/models/ir_http.py
Comment (xmo-odoo) [inline addons/website/models/ir_http.py:136]: Although it's mostly been dropped as a concept, maybe the [DNT header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/DNT) could be used instead? It has the advantage of well-defined semantics, and respecting it for the 3 users which do use it is nice.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (xmo-odoo) [inline addons/website_event_track/static/src/js/service_worker.js:157]: `Headers` objects are absolutely not required, the `headers` key can just be an object literal (and so this function seems pretty useless).

**Review by xmo-odoo — COMMENTED** [review]
@robodoo override=ci/security

Not actually touching the existing getattr code.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js]: Hello @pparidans, just to ping you about my question above. Thanks.

File: addons/website/models/ir_http.py
Comment (pparidans) [reply addons/website/models/ir_http.py]: I'm a bit confused about that rename...
The Do Not Track header exists (even if now deprecated) with a similar meaning... but a different purpose (avoid tracking from the user point of view, not the server).
Wouldn't it be confusing to use a similar name?

Ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/DNT

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js]: Thanks for the explanation, I will left it without the header as suggested as it is concerning only one URL.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js]: Yes thanks for the tip, I have inlined the function in the code (as there is only one use now) and reduced it to one line using the object literal as parameter as suggested.

File: addons/website/models/ir_http.py
Comment (pyduf) [reply addons/website/models/ir_http.py]: I think that X-Do-Not-Track is ok because it is a custom header (starting with X-) and it is different than DNT.
And in either case, it is the client that informs the server to not track the page (here the server cannot know that the client is prefetching the page).

But indeed, it could be confused with the original meaning of DNT which is an user preference and not added for a technical reason (here prefetch).

So maybe one of the proposition of rdeodoo or maybe: X-Disable-Tracking

Let me know what you prefer. Thanks

Comment (rdeodoo) [reply addons/website/models/ir_http.py]: Anything will do for me :)
I like your proposal tho

Comment (pparidans) [reply addons/website/models/ir_http.py]: Like it too :)

Comment (pyduf) [reply addons/website/models/ir_http.py]: Thanks, I have then used X-Disable-Tracking.
I have also changed the condition slightly as 
request.httprequest.headers.get('X-Disable-Tracking') == '1'

(before the presence of the header alone was disabling the tracking, for example '0' is True)

**Review by pparidans — APPROVED** [review]
LGTM :+1:

I leave the final word to @tde-banana-odoo & @rdeodoo about the backporting topic

**Comment by tde-banana-odoo** [comment]
Thanks for your input :) it will anyway go through the testing / final review process, then we will decide if we backport of not at merge time.

File: addons/website/models/ir_http.py
Comment (awa-odoo) [inline addons/website/models/ir_http.py:136]: Would be nice to have this change isolated in its own commit with an explanation that we introduce it to make it possible to skip the tracking of certain pages / use cases (you can give the service worker example).

File: addons/test_event_full/tests/test_performance.py
Comment (awa-odoo) [inline addons/test_event_full/tests/test_performance.py:447]: Can I ask why this adds 10 and then 9 requests?
Is it because of the added `track="1"` on the template alone?

If so, would be great to explain that in the commit message.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/service_worker.js:247]: Do we want to track at all in the service worker?

We could say that we only track regular page accesses and simplify it here by always adding the disable tracking header.

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Added some comments / questions.
Don't hesitate to tell me if anything is unclear or if I misunderstood something.

Thanks & cheers!

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js:247]: That would be nice indeed. What do you think @pparidans ? If I have well understood your previous message, the requests in processPendingRequests are real call from the user (replayed) but you also mentioned that it concerns only one URL.

File: addons/test_event_full/tests/test_performance.py
Comment (pyduf) [reply addons/test_event_full/tests/test_performance.py:447]: Actually, that was the question I was asking in the task because when running the test locally I get crazy number (10 times higher) :-)

As the only thing we do here is to add an header that can potentially decrease the number of queries (no tracking for some pages) and add the track="1" on event that add some queries (to track the page: visitor, ...).
I think we can safely assume that the added queries are due to the tracking of that additional page.
I have just taken the number from the runbot report.
I will add a comment in the commit explaining the added page to track increase the number of queries in those tests.

I have investigate why I don't get the same value locally as in the runbot but failed until now. Here is what I have found:

The 2 tests are annotated with the warmup decorator which run test first without checking the assertion and then a second time, this time checking the assertion.
The decorator restore the DB between the 2 tests.
I suppose the idea is that the second time there will be no prefetch of the URL in the page because the browser cache will be filled.
It works on the runbot but not locally. It is like the browser cache is cleaned between the 2 run locally and not in the runbot but I don't see why (library version ? No demo data ?).

If I do manually what warmup do but without restoring DB, invalidate cache, ..., I get similar result as in the runbot:

test_events_browse_event_manager
First time: 475
Second time: 56

test_events_browse_public
First time: 359
Second time: 39

File: addons/website/models/ir_http.py
Comment (pyduf) [reply addons/website/models/ir_http.py:136]: Good idea, it is done.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pparidans) [reply addons/website_event_track/static/src/js/service_worker.js:247]: > Do we want to track at all in the service worker?
> We could say that we only track regular page accesses and simplify it here by always adding the disable tracking header.
Bear in mind that once you have a service worker which listen to "fetch" event, **all** requests pass through it! So disabling tracking for all fetch request would result in no tracking at all (at least under `/event` in this case).

File: addons/website_event_track/static/src/js/service_worker.js
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/service_worker.js:316]: I'm sorry but I'm still not sure about this @pparidans @pyduf

So this is the call we want to track, since it has the "true" argument.
If I grep in the code, that method is called when the action "prefetch-pages" is called.
But at that point, the user has already "rendered" the event page, and we have tracked that call, correct?

What you mean is, once the PWA is installed, we would not track anything anymore if we always add the header here.
That seems ok to me, as the PWA usage is very niche anyway.

Please tell me if I misunderstood anything.

Comment (pparidans) [reply addons/website_event_track/static/src/js/service_worker.js:316]: Don't forget that, once installed (cf. after the very first page request), a service worker actually works with a "scope" (ie. `/event`) and not a a single page. So every request made under that scope will pass through the service worker's fetch event listener.

In other words, disabling tracking globally in the service worker will then disable tracking for every request made to a subpath of `/event` (aka. the scope)!

Comment (pparidans) [reply addons/website_event_track/static/src/js/service_worker.js:316]: Actually "installing the app" has nothing to do with the service worker behavior: he'll be present and catching requests anyway.

Having a service worker and a (quite minimal) handling of offline are "only" some of the prerequisite for a website to be seen as installable by the browser.

Comment (awa-odoo) [inline addons/website_event_track/static/src/js/service_worker.js:316]: Could we change to `await processFetchRequest(new Request(url), { disableTracking: true });` ?

That way:
- it's much clearer when calling the method that we disable the tracking
- we name the parameter closer to the header name

Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js]: Yes indeed, that's easier to read. It's done.

File: addons/test_event_full/tests/test_performance.py
Comment (pyduf) [reply addons/test_event_full/tests/test_performance.py:447]: As suggested, I have pushed the code without track=1 for the "event" page and the rest unchanged and I get the same number as before so it shows that  he added queries come from the added track=1 attribute for the "event" page.

Without track=1 on the page /event:
test_events_browse_event_manager: # com runbot: 49 - ent runbot:50
test_events_browse_public: # com runbot: 30 - ent runbot: 30

File: addons/website_event_track/static/src/js/service_worker.js
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/service_worker.js:241]: I think the syntax for those should be as following:

```js
/**
  * @param {Object} [options]
  * @param {Boolean} [options.disableTracking] blabla
*/
```

File: addons/website_event_track/static/src/js/service_worker.js
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/service_worker.js:249]: `if (options && options.disableTracking)` to allow calling the function without proving options since you marked it as optional in the docstring?

File: addons/website_event_track/static/src/js/service_worker.js
Comment (awa-odoo) [inline addons/website_event_track/static/src/js/service_worker.js:357]: I'm surprised this call works with your condition here above?

image

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Almost there, some extra polishing comments.

Thanks & cheers!

File: addons/test_event_full/tests/test_performance.py
Comment (awa-odoo) [reply addons/test_event_full/tests/test_performance.py:447]: Ok so we know it's because of that change then, thanks for checking in-depth!

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js:249]: Actually, I had added the following statement at the beginning of the function:
options = options || {};

But as there is only one argument, I have used your solution.

File: addons/website_event_track/static/src/js/service_worker.js
Comment (pyduf) [reply addons/website_event_track/static/src/js/service_worker.js:241]: Thanks, I didn't know. It is changed.

**Comment by awa-odoo** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**
